include/execution.hpp

```diff
@@ -84,15 +96,16 @@ namespace std::execution {
 
   template<class S>
   auto __sender_traits_base_fn() {
+    using P = __connect_awaitable_::__promise_t<>;
     if constexpr (__has_sender_types<S>) {
       return __typed_sender<S>{};
     } else if constexpr (derived_from<S, sender_base>) {
       return sender_base{};
-    } else if constexpr (__awaitable<S>) { // NOT TO SPEC
-      if constexpr (is_void_v<__await_result_t<S>>) {
+    } else if constexpr (__awaitable<S, P>) {
```

Comment:
We don't have a receiver here to test with, so we test with a fake receiver, `__test_receiver`. This is essentially asking the question: can I await this thing in a coroutine whose promise type implements `await_transform` to dispatch to `as_awaitable`? It will try to dispatch to `as_awaitable` with essentially a bogus promise type. If some type has customized `as_awaitable` and is relying on being passed a promise type with specific properties (e.g., certain receiver queries), then this could conceivably cause strange errors. The fix would be to specialize `sender_traits` for that type.

The alternative would be to only consider types that are awaitable in coroutines whose promise typed do _not_ implement `await_transform`. A type like `task<>` would have to provide the typed sender traits manually in that case, which is a lot of scary-looking boilerplate at the moment.

This is the part of this PR I'm least comfortable with. I'm curious what others think.

Reply:
I don't think we can, in general, know what type the user is going to get back when they await something.

We can only really just rely on `__awaitable` concept / interface if we know that there is no `as_awaitable()` overload for the concrete promise type that ends up being generated by the `connect()` implementation.
But since there could be a constrained `as_awaitable()` implementation, probing its existence with a dummy promise-type might miss that fact.

e.g. say we have a type that is used to extract the current allocator from the coroutine
```
struct get_current_allocator_t {
  template<typename Promise>
    requires allocator_provider<Promise>
  auto tag_invoke(as_awaitable_t, get_current_allocator_t, Promise& p) noexcept {
    struct awaitable {
      Promise& p;
      bool await_ready() noexcept { return true; }
      void await_suspend(coroutine_handle<>) noexcept { std::terminate(); }
      auto await_resume() noexcept { return get_allocator(p); }
    };
    return awaitable{p};
  }
};
inline constexpr get_current_allocator_t get_current_allocator;
```

Which can be used within a coroutine to get the current allocator object:
```
some_task<T> example() {
  auto alloc = co_await get_current_allocator;
}
```

If you ask if such a type is `__awaitable` in a coroutine with promise `__connect_awaitable::__promise_t<>` then this would return true and this implementation would give it a sender_traits that indicated it returned, say `std::allocator` if it picked up the default implementation - or possibly `void` if the above `__test_receiver` was used.

However, if you try to use it within a concrete task type or use it within a sender then it might return a different type.

For the above type, it really shouldn't satisfy the `typed_sender` concept since it doesn't have a single result type, but this implementation would still potentially treat it as if it were a sender.

This is one of the reasons why the implementation of awaitables-as-senders requires the awaitable to manually implement the sender-traits in order to be able to support the sender concept. We can't always reliable deduce the result-type of an awaitable outside of the context of a particular coroutine type due to the nature of `await_transform`/`as_awaitable`.